Remove lodash from login selectors

sgomes opened the pull request:
Removing `lodash` from state, particularly `_.get`, will require some level of understanding of each portion of state, so that we avoid falling into the trap of using `?.` everywhere instead of just where it's needed.

I prepared this small PR to get some feedback on my approach, and whether you think it's the right one.

For example, for `login` state, I decided that selectors shouldn't need to able to work with `undefined` state, but rather can expect for state to have been initialised by the base reducer, such that `state.login` is always present and is an object. Is this the right approach?

#### Changes proposed in this Pull Request

* Remove `lodash`, particularly `_.get`, from the login selectors

#### Testing instructions

* Ensure that the login selector tests pass
* Smoke-test login and ensure it continues to work correctly

sgomes committed:
Remove lodash from login selectors

sgomes commented on client/state/login/test/selectors.js:
```diff
@@ -28,10 +28,14 @@ import {
 	getSocialAccountLinkService,
 } from '../selectors';
 
+const EMPTY_STATE = {
```
Is it a reasonable expectation for these tests that state should always have a `login` key, and it is initialised to an object?

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -58,7 +52,7 @@ export const getTwoFactorPushToken = ( state ) =>
  */
 export const getTwoFactorPushPollInProgress = ( state ) =>
-	get( state, 'login.twoFactorAuthPushPoll.inProgress', false );
+	state.login.twoFactorAuthPushPoll?.inProgress ?? false;
```
Unlike the `twoFactorAuth` state, which can be `null`, `twoFactorAuthPushPoll` is always a valid object where both its props are always valid booleans. The selector can be the simplest possible property access without any optional chaining or defaults.

sgomes replied:
I agree that this can and should be simplified, good catch! 👍 

However, this does mean that we now need to add this default state to the tests, or otherwise they will fail.

@tyxla @jsnajdr: What do you think is the correct approach?
- Make the selector tests depend on the reducer, so that we can use it to prepare the empty state that we test against?
- Keep the selector tests independent from the reducer, and hardcode the default values into the empty state in the test, e.g.:
  ```js
  const EMPTY_STATE = {
    login: {
      twoFactorAuthPushPoll: {
        inProgress: false,
        success: false,
      },
    },
  };
  ```

tyxla replied:
> However, this does mean that we now need to add this default state to the tests,

I'd say yes. Tests are supposed to mirror the original execution flow, and if we need to be more specific when defining test fixtures, we should go that way.

> Make the selector tests depend on the reducer, so that we can use it to prepare the empty state that we test against?

I'd pick that one. I don't have any reservations against this approach, and have done it before without problems.

jsnajdr replied:
> Make the selector tests depend on the reducer, so that we can use it to prepare the empty state that we test against?

When reviewing the selectors and deciding whether a particular optional chaining or fallback is needed, I used the reducer code as the source of truth for that. Sometimes the actions are relevant, too. That means they are all part and parcel of one module and it's OK to test them as one unit.

When writing new Redux-related tests, I usually try to create a simple Redux store with the tested reducer and use the actions and selectors on this store. Then it feels that the unit tests do something that's close to a real-world app.

sgomes replied:
Thank you both for the feedback! 👍 

I went with the simpler approach of calling the reducer with an `undefined` previous state and a no-op action. It's not as extensive as creating a store, but I think it fits the model of the selector tests well. Let me know if you disagree!

jsnajdr replied:
> I went with the simpler approach of calling the reducer with an undefined previous state and a no-op action.

I think that's perfectly fine. By calling `newState = reducer( oldState, action )`, you implemented almost the entire Redux (without subscriptions) on one line 🙂 

The point is that you can create realistic test states just by calling the reducer with an action, instead of maintaining large testing objects and having to modify them on every state shape change.

sgomes commented:
Re: the matticbot results, they're not entirely unexpected. Due to limitations in our current tooling, we're still transpiling `?.` and `??` syntax, which results in some amount of inline code. It does compress well, however.

Once we're on `webpack` 5 and we remove the pinned Safari 13 requirement, we should get the new syntax in the evergreen build, and a nice reduction in uncompressed bundle size as a result.

sgomes commented:
Note: E2E failures appear to be unrelated, as they're failing in other PRs as well.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -76,9 +70,9 @@ export const getTwoFactorPushPollSuccess = ( state ) =>
+	const twoFactorAuth = state.login.twoFactorAuth;
 
-	return ! isEmpty( twoFactorAuth );
+	return !! ( twoFactorAuth && Object.keys( twoFactorAuth ).length );
```
Seems to me this could be a simple `state.login.twoFactorAuth != null` check. If you look at the reducer, it will never set the state to an empty object: it does the `isEmpty` check itself and sets the state to `null` if the object turns out to be empty.

sgomes replied:
Agreed; good idea! Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -88,7 +82,7 @@ export const isTwoFactorEnabled = ( state ) => {
  */
 export const isRequestingTwoFactorAuth = ( state ) => {
-	return get( state, 'login.isRequestingTwoFactorAuth', false );
+	return state.login.isRequestingTwoFactorAuth ?? false;
```
Here the state is always a valid boolean and doesn't need a fallback.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -98,7 +92,7 @@ export const isRequestingTwoFactorAuth = ( state ) => {
  */
 export const getTwoFactorAuthRequestError = ( state ) => {
-	return get( state, 'login.twoFactorAuthRequestError', null );
+	return state.login.twoFactorAuthRequestError ?? null;
```
Also shouldn't need a fallback. The state is either `null` or an `error` object. The only case where the fallback would kick in is if the `TWO_FACTOR_AUTHENTICATION_LOGIN_REQUEST_FAILURE` action object had `undefined` `action.error` property.

sgomes replied:
Good point. If that situation is relevant and it should indeed be a `null`, then that should be handled once by the reducer, not every time by the selector.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -132,7 +126,7 @@ export const isTwoFactorAuthTypeSupported = ( state, type ) => {
  */
 export const isRequesting = ( state ) => {
-	return get( state, 'login.isRequesting', false );
+	return state.login.isRequesting ?? false;
```
No need for fallback, always a valid boolean.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -142,7 +136,7 @@ export const isRequesting = ( state ) => {
  */
 export const getRequestError = ( state ) => {
-	return get( state, 'login.requestError', null );
+	return state.login.requestError ?? null;
```
Again, shouldn't need a fallback. Except if the `*_FAILURE` action objects had an empty `action.error` field...

sgomes replied:
Agreed, done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -152,7 +146,7 @@ export const getRequestError = ( state ) => {
  */
 export const getRequestNotice = ( state ) => {
-	return get( state, 'login.requestNotice', null );
+	return state.login.requestNotice ?? null;
```
Similar case as with the error states: `requestNotice` is either `null` or a valid object, except if `action.notice` was undefined in relevant actions.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -164,7 +158,7 @@ export const getRequestNotice = ( state ) => {
  */
 export const getRedirectToOriginal = ( state ) => {
-	return get( state, 'login.redirectTo.original', null );
+	return state.login.redirectTo?.original ?? null;
```
The `redirectTo` reducer is a result of `combineReducers`, is always a valid object, and always has the `.original` and `.sanitize` properties. No need to optionally chain `redirectTo?.original`.

sgomes replied:
Good point, `combineReducers` does mean it'll always be available in this case. Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -185,7 +179,7 @@ export const getRedirectToSanitized = ( state ) => {
  */
 export const isFormDisabled = ( state ) => {
-	return get( state, 'login.isFormDisabled', false );
+	return state.login.isFormDisabled ?? false;
```
Always a valid bool.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -195,7 +189,7 @@ export const isFormDisabled = ( state ) => {
  */
 export const getAuthAccountType = ( state ) => {
-	return get( state, 'login.authAccountType', null );
+	return state.login.authAccountType ?? null;
```
No need to fallback. The state is either `null` or the `action.data.type` field of a REST response.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -214,7 +207,7 @@ export const isSocialAccountCreating = ( state ) =>
  */
 export const getCreatedSocialAccountUsername = ( state ) =>
-	get( state, 'login.socialAccount.username', null );
+	state.login.socialAccount?.username ?? null;
```
`socialAccount` always contains a valid object. It just sometimes doesn't have the `username` property. That means the optional chaining is not needed, and the fallback is needed.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -204,8 +198,7 @@ export const getAuthAccountType = ( state ) => {
  */
-export const isSocialAccountCreating = ( state ) =>
-	get( state, 'login.socialAccount.isCreating', null );
+export const isSocialAccountCreating = ( state ) => state.login.socialAccount?.isCreating ?? null;
```
No need for optional chaining nor the fallback. `socialAccount` is always present and always has the `isCreating` property, which is always a boolean.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -241,7 +234,7 @@ export const getCreateSocialAccountError = ( state ) =>
  */
 export const getRequestSocialAccountError = ( state ) =>
-	get( state, 'login.socialAccount.requestError', null );
+	state.login.socialAccount?.requestError ?? null;
```
For `bearerToken`, `createError` and `requestError`, the optional chaining is redundant and the fallback is OK.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -250,16 +243,15 @@ export const getRequestSocialAccountError = ( state ) =>
  */
 export const getSocialAccountIsLinking = ( state ) =>
-	get( state, 'login.socialAccountLink.isLinking', null );
+	state.login.socialAccountLink?.isLinking ?? null;
```
Both optional chaining and the fallback can be removed. Always a valid object and a valid boolean, respectively.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -250,16 +243,15 @@ export const getRequestSocialAccountError = ( state ) =>
  */
-export const getSocialAccountLinkEmail = ( state ) =>
-	get( state, 'login.socialAccountLink.email', null );
+export const getSocialAccountLinkEmail = ( state ) => state.login.socialAccountLink?.email ?? null;
```
Optional chaining not needed, fallback is OK.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -268,7 +260,7 @@ export const getSocialAccountLinkEmail = ( state ) =>
  */
 export const getSocialAccountLinkService = ( state ) =>
-	get( state, 'login.socialAccountLink.authInfo.service', null );
+	state.login.socialAccountLink?.authInfo?.service ?? null;
```
The optional chain op after `socialAccountLink` is not needed, the rest is OK.

sgomes replied:
Done.

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -286,4 +278,4 @@ export const getSocialAccountLinkAuthInfo = ( state ) =>
  */
 export const getLastCheckedUsernameOrEmail = ( state ) =>
-	get( state, 'login.lastCheckedUsernameOrEmail', null );
+	state.login.lastCheckedUsernameOrEmail ?? null;
```
No need to fallback.

sgomes replied:
Done.

jsnajdr commented on client/state/login/test/selectors.js:
```diff
@@ -28,10 +28,14 @@ import {
-			const id = getTwoFactorUserId( undefined );
+			const id = getTwoFactorUserId( EMPTY_STATE );
 
 			expect( id ).to.be.null;
```
If you don't mind more labor-intensive changes, it would be awesome to migrate these Chai asserts to Jest 🙂

sgomes replied:
Done! 🙂

jsnajdr reviewed (commented):
I think it's ok to assume that `state.login` is always there. In practice, it will be. Posted a lot of comments about redundancy of some optional chainings and fallbacks.

sgomes committed:
Switch tests from chai to jest

tyxla reviewed (commented):
Not much to add to Jarda's extensive code review here already, other than the fact that this is working great in testing 👍

sgomes committed:
Simplify isTwoFactorEnabled

sgomes committed:
Simplify getTwoFactorAuthRequestError

sgomes committed:
Use reducer to initialise empty state

sgomes committed:
Simplify twoFactorAuthPushPoll selectors

sgomes committed:
Simplify isRequestingTwoFactorAuth

sgomes committed:
Simplify isRequesting

sgomes committed:
Simplify getRequestError

sgomes committed:
Simplify getRequestNotice

sgomes committed:
Simplify getRedirectTo*

sgomes committed:
Simplify isFormDisabled

sgomes committed:
Simplify getAuthAccountType

sgomes committed:
Simplify socialAccount selectors

sgomes committed:
Simplify socialAccountLink selectors

sgomes committed:
Simplify getLastCheckedUsernameOrEmail

sgomes committed:
Use arrow function shorthand syntax where possible

sgomes commented:
Thank you for the extensive review, @jsnajdr and @tyxla! 🙇
This is exactly what I was hoping for in an initial attempt. If anything, I should have picked a smaller portion of state! 😄 

I've addressed all comments, and here's my takeaway from the discussion, to be applied whenever I'm revamping selectors:
- Don't just remove `lodash`; analyse every fallback and part of the chain to determine what's really needed, and use as few optional chains and fallbacks as possible.
- Prefer simplicity in the selectors over simplicity in the tests or reducer
- It's okay to depend on the reducer in the selector tests to provide an initial/empty state

I think I've addressed everything in code as well, but please let me know if I missed anything!

jsnajdr commented on client/state/login/selectors.js:
```diff
@@ -27,9 +20,8 @@ export const getTwoFactorUserId = ( state ) => {
-	return state.login.twoFactorAuth[ `two_step_nonce_${ nonceType }` ] ?? null;
-};
+export const getTwoFactorAuthNonce = ( state, nonceType ) =>
+	state.login.twoFactorAuth[ `two_step_nonce_${ nonceType }` ] ?? null;
```
Looks like there's a pre-existing bug here: `twoFactorAuth` can be `null`, so the property access should be optional: ``twoFactorAuth?.[ `two_step_...` ]``

sgomes replied:
Good catch, done!

jsnajdr commented on client/state/login/test/selectors.js:
```diff
@@ -45,15 +50,15 @@ describe( 'selectors', () => {
 			} );
 
-			expect( id ).to.equal( 123456 );
+			expect( id ).toStrictEqual( 123456 );
```
Nit: this check can be a simple `.toBe`. [`toStrictEqual`](https://jestjs.io/docs/en/expect.html#tostrictequalvalue) is an extended version of deep equality check, useful for checking class instances. I don't remember ever seeing it in practice. We'd probably use it more if there was more OOP and prototypes in our code.

sgomes replied:
I should probably have read the Jest documentation a bit closer. Thanks, done!

jsnajdr commented on client/state/login/test/selectors.js:
```diff
@@ -83,13 +88,13 @@ describe( 'selectors', () => {
 	describe( 'isRequestingTwoFactorAuth', () => {
 		test( 'should return false by default', () => {
-			expect( isRequestingTwoFactorAuth( undefined ) ).to.be.false;
+			expect( isRequestingTwoFactorAuth( EMPTY_STATE ) ).toBeFalsy();
```
`.toBeFalsy()` also accepts `undefined`, `null` or `0`. We can be more strict here and test explicitly for a boolean `.toBe( false )`. That applies to most (if not all) other usages of `.toBe{Falsy,Truthy}` in this test, too.

sgomes replied:
Good point, we probably do want to be checking for the stricter values, because reading the Chai docs that's what the previous code did.

This helped find a bug in the `isFormDisabled` reducer, which initialised as `null` instead of `false`!

jsnajdr commented on client/state/login/test/selectors.js:
```diff
@@ -364,7 +369,7 @@ describe( 'selectors', () => {
 					},
 				} )
-			).to.eql( true );
+			).toEqual( true );
```
Nit: primitive values like numbers, booleans and strings can be compared with `.toBe`. `.toEqual` is usually used only to deep-compare arrays and objects.

sgomes replied:
Replaced all those with `toBe` 👍

jsnajdr reviewed (commented):
Found one bug and posted a few style nits about Jest asserts 🙂

sgomes committed:
Fix potential bug in getTwoFactorAuthNonce

sgomes committed:
Bugfix: isFormDisabled initialising as null

sgomes committed:
Apply review suggestions to selector tests

jsnajdr reviewed (approved):
Nice improvement that's now ready to merge :+1: It's funny that the bundle size is actually bigger now 🙂 Hopefully it moves back in the right direction after webpack and Terser upgrade.

sgomes commented:
Woohoo, thank you for the extremely thorough and helpful review, @jsnajdr! 👍 This PR not only removes `lodash` now, but it simplifies the selectors and the tests, and it even fixes a few bugs 🙂
